server/services/v1/query_runner.go

```diff
@@ -150,62 +153,129 @@ func (runner *BaseQueryRunner) GetCollections(db *metadata.Database, collName st
 	return collection, nil
 }
 
-func (runner *BaseQueryRunner) extractIndexParts(userDefinedKeys []*schema.Field, doc []byte) ([]interface{}, error) {
+// extractIndexParts extract the keys out from the document and return the following
+//  - []byte: the document is returned because it may be modified if autoGenerated is set on any field
+//  - []interface - a slice of values for the primary keys
+//  - []byte - this is returned because we need to return the keys of all the rows
+//  - bool - if we need to call insert instead of Replace(for autoGenerate special cases)
+//  - error - if any error occurred
+func (runner *BaseQueryRunner) extractIndexParts(ctx context.Context, table []byte, userDefinedKeys []*schema.Field, doc []byte) ([]byte, []interface{}, []byte, bool, error) {
 	var appendTo []interface{}
+	var keysReturnedInResp []byte
+	var autoGeneratedTypes []schema.FieldType
 	for _, v := range userDefinedKeys {
-		jsonVal, _, _, err := jsonparser.Get(doc, v.FieldName)
+		jsonVal, dtp, _, err := jsonparser.Get(doc, v.FieldName)
 		if err != nil {
-			return nil, api.Errorf(codes.InvalidArgument, errors.Wrapf(err, "missing index key column(s) '%s'", v.FieldName).Error())
+			if dtp != jsonparser.NotExist || !v.IsAutoGenerated() {
+				return nil, nil, nil, false, api.Errorf(codes.InvalidArgument, errors.Wrapf(err, "parsing index key column(s) '%s'", v.FieldName).Error())
+			}
+			if dtp == jsonparser.NotExist && !v.IsAutoGenerated() {
```

Comment:
This need to be moved inside above `if` and be just `if !v.IsAutoGenerated`.

At this point `dtp == jsonparser.NotExist && v.IsAutoGenerated` is true, so the condition is always false.